File: bench/python/ops/bench_composite.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""CV-CUDA Composite operator benchmark - Python equivalent of BenchComposite.cpp"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import cvcuda  # noqa: E402
from python_bench_utils import (  # noqa: E402
    get_input_kind,
    parse_shape,
    get_dtype,
    get_dtype_size,
    get_num_channels,
    get_format_from_dtype,
    create_tensor,
    create_image_batch_varshape,
    create_stream_cache,
    run_benchmark,
)


def composite(state):
    """Composite operator benchmark matching C++ BenchComposite.cpp"""

    N, H, W = parse_shape(state.get_string("shape"))
    dtype_str = state.get_string("InOutDataType")
    dtype = get_dtype(dtype_str)
    input_kind = get_input_kind(state.get_string("inputKind"))
    out_channels = state.get_int64("outChannels")
    try:
        layout = state.get_string("layout")
    except (KeyError, RuntimeError):
        layout = "NHWC"
    device_id = state.get_device()

    num_channels = get_num_channels(dtype_str)

    if out_channels not in (3, 4):
        state.skip(f"Invalid outChannels: {out_channels}")
        return None
    if layout not in ("NHWC", "NCHW", "NCHW_FAKE"):
        state.skip(
            "Composite benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts"
        )
        return None
    is_planar = layout == "NCHW"
    is_fake_planar = layout == "NCHW_FAKE"
    if is_fake_planar and input_kind == "VarShape":
        state.skip("Fake-planar (NCHW_FAKE) Composite benchmark is tensor-only")
        return None

    dtype_size = get_dtype_size(dtype_str)
    base_dtype_size = get_dtype_size(dtype)
    mask_dtype_size = 1  # uint8 mask
    state.add_global_memory_reads(N * H * W * (dtype_size * 2 + mask_dtype_size))
    state.add_global_memory_writes(N * H * W * out_channels * base_dtype_size)

    if is_fake_planar:
        fg = create_tensor(
            (N, num_channels, H, W),
            dtype,
            device_id,
            layout="NCHW",
            fill_mode="checkerboard",
        )
        bg = create_tensor(
            (N, num_channels, H, W),
            dtype,
            device_id,
            layout="NCHW",
            fill_mode="checkerboard",
        )
        mask = create_tensor(
            (N, 1, H, W),
            cvcuda.Type.U8,
            device_id,
            layout="NCHW",
            fill_mode=1,
        )
        inter_fg = create_tensor(
            (N, H, W, num_channels),
            dtype,
            device_id,
            layout="NHWC",
            fill_mode=0,
        )
        inter_bg = create_tensor(
            (N, H, W, num_channels),
            dtype,
            device_id,
            layout="NHWC",
            fill_mode=0,
        )
        inter_mask = create_tensor(
            (N, H, W, 1),
            cvcuda.Type.U8,
            device_id,
            layout="NHWC",
            fill_mode=0,
        )
        inter_dst = create_tensor(
            (N, H, W, out_channels),
            dtype,
            device_id,
            layout="NHWC",
            fill_mode=0,
        )
        dst = create_tensor(
            (N, out_channels, H, W),
            dtype,
            device_id,
            layout="NCHW",
            fill_mode=0,
        )
    elif input_kind == "Tensor":  # Tensor mode
        tensor_shape = (N, num_channels, H, W) if is_planar else (N, H, W, num_channels)
        mask_shape = (N, 1, H, W) if is_planar else (N, H, W, 1)
        dst_shape = (N, out_channels, H, W) if is_planar else (N, H, W, out_channels)
        fg = create_tensor(
            tensor_shape,
            dtype,
            device_id,
            layout=layout,
            fill_mode="checkerboard",
        )
        bg = create_tensor(
            tensor_shape,
            dtype,
            device_id,
            layout=layout,
            fill_mode="checkerboard",
        )
        mask = create_tensor(
            mask_shape,
            cvcuda.Type.U8,
            device_id,
            layout=layout,
            fill_mode=1,
        )
        dst = create_tensor(
            dst_shape,
            dtype,
            device_id,
            layout=layout,
            fill_mode=0,
        )
    else:  # ImageBatchVarShape mode
        img_format = get_format_from_dtype(dtype_str, num_channels, planar=is_planar)
        dst_format = get_format_from_dtype(dtype_str, out_channels, planar=is_planar)
        mask_format = cvcuda.Format.U8
        fg = create_image_batch_varshape(
            (N, H, W, num_channels),
            0,
            img_format,
            dtype,
            device_id,
            fill_mode="checkerboard",
        )
        bg = create_image_batch_varshape(
            (N, H, W, num_channels),
            0,
            img_format,
            dtype,
            device_id,
            fill_mode="checkerboard",
        )
        mask = create_image_batch_varshape(
            (N, H, W), 0, mask_format, cvcuda.Type.U8, device_id, fill_mode=1
        )
        dst = create_image_batch_varshape(
            (N, H, W, out_channels),
            0,
            dst_format,
            dtype,
            device_id,
            fill_mode="checkerboard",
        )

    get_stream = create_stream_cache()

    def run(launch):
        stream = get_stream(launch)
        if is_fake_planar:
            cvcuda.reformat_into(inter_fg, fg, stream=stream)
            cvcuda.reformat_into(inter_bg, bg, stream=stream)
            cvcuda.reformat_into(inter_mask, mask, stream=stream)
            cvcuda.composite_into(
                inter_dst, inter_fg, inter_bg, inter_mask, stream=stream
            )
            cvcuda.reformat_into(dst, inter_dst, stream=stream)
        else:
            cvcuda.composite_into(dst, fg, bg, mask, stream=stream)

    return run


if __name__ == "__main__":
    run_benchmark("composite", composite)

File: bench/cpp/ops/BenchComposite.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../CppBenchUtils.hpp"
#include "ops/generated/BenchCompositeConfig.hpp"

#include <cvcuda/OpComposite.hpp>
#include <cvcuda/OpReformat.hpp>

#include <nvbench/nvbench.cuh>

#include <stdexcept>
#include <string>
#include <string_view>

enum class CompositeLayout
{
    NHWC,
    NCHW,
    NCHW_FAKE,
    INVALID
};

inline bool IsPlanar(CompositeLayout layout)
{
    return layout == CompositeLayout::NCHW;
}

inline bool IsFakePlanar(CompositeLayout layout)
{
    return layout == CompositeLayout::NCHW_FAKE;
}

inline void ValidateOutChannels(int outChannels)
{
    if (outChannels == 3 || outChannels == 4)
    {
        return;
    }

    throw std::invalid_argument("Invalid outChannels = " + std::to_string(outChannels));
}

inline CompositeLayout ParseCompositeLayout(nvbench::state &state, std::string_view layout)
{
    if (layout == "NHWC")
    {
        return CompositeLayout::NHWC;
    }
    if (layout == "NCHW")
    {
        return CompositeLayout::NCHW;
    }
    if (layout == "NCHW_FAKE")
    {
        return CompositeLayout::NCHW_FAKE;
    }

    state.skip("Composite benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts");
    return CompositeLayout::INVALID;
}

inline CompositeLayout ValidateCompositeBenchParams(nvbench::state &state, std::string_view layout, int outChannels)
{
    ValidateOutChannels(outChannels);
    return ParseCompositeLayout(state, layout);
}

inline bool SkipUnsupportedCompositeCase(nvbench::state &state, benchutils::InputKind inputKind, CompositeLayout layout)
{
    if (IsFakePlanar(layout) && inputKind == benchutils::InputKind::VarShape)
    {
        state.skip("Fake-planar (NCHW_FAKE) Composite benchmark is tensor-only");
        return true;
    }

    return false;
}

template<typename T, typename M>
void RunFakePlanarComposite(nvbench::state &state, cvcuda::Composite &op, int3 shape, int outChannels)
{
    using BT = typename nvcv::cuda::BaseType<T>;

    const int ch = nvcv::cuda::NumElements<T>;

    nvcv::Tensor fg(
        {
            {shape.x, ch, shape.y, shape.z},
            "NCHW"
    },
        benchutils::GetDataType<BT>());
    nvcv::Tensor bg(
        {
            {shape.x, ch, shape.y, shape.z},
            "NCHW"
    },
        benchutils::GetDataType<BT>());
    nvcv::Tensor mask(
        {
            {shape.x, 1, shape.y, shape.z},
            "NCHW"
    },
        benchutils::GetDataType<M>());
    nvcv::Tensor interFg(
        {
            {shape.x, shape.y, shape.z, ch},
            "NHWC"
    },
        benchutils::GetDataType<BT>());
    nvcv::Tensor interBg(
        {
            {shape.x, shape.y, shape.z, ch},
            "NHWC"
    },
        benchutils::GetDataType<BT>());
    nvcv::Tensor interMask(
        {
            {shape.x, shape.y, shape.z, 1},
            "NHWC"
    },
        benchutils::GetDataType<M>());
    nvcv::Tensor interDst(
        {
            {shape.x, shape.y, shape.z, outChannels},
            "NHWC"
    },
        benchutils::GetDataType<BT>());
    nvcv::Tensor dst(
        {
            {shape.x, outChannels, shape.y, shape.z},
            "NCHW"
    },
        benchutils::GetDataType<BT>());

    benchutils::FillTensor<BT>(fg, benchutils::CheckerboardValues<BT>());
    benchutils::FillTensor<BT>(bg, benchutils::CheckerboardValues<BT>());
    benchutils::FillTensor<M>(mask, [](const long4_16a &) { return 1; });

    cvcuda::Reformat reformatOp;

    benchutils::warmup_and_exec(
        state, BENCH_COMPOSITE_WARMUP_ITERATIONS,
        [&op, &reformatOp, &fg, &bg, &mask, &interFg, &interBg, &interMask, &interDst, &dst](cudaStream_t s)
        {
            reformatOp(s, fg, interFg);
            reformatOp(s, bg, interBg);
            reformatOp(s, mask, interMask);
            op(s, interFg, interBg, interMask, interDst);
            reformatOp(s, interDst, dst);
        });
}

template<typename BT>
nvcv::Tensor MakeCompositeTensor(int3 shape, int channels, bool isPlanar)
{
    if (isPlanar)
    {
        return nvcv::Tensor(
            {
                {shape.x, channels, shape.y, shape.z},
                "NCHW"
        },
            benchutils::GetDataType<BT>());
    }

    return nvcv::Tensor(
        {
            {shape.x, shape.y, shape.z, channels},
            "NHWC"
    },
        benchutils::GetDataType<BT>());
}

template<typename T, typename M>
void RunTensorComposite(nvbench::state &state, cvcuda::Composite &op, int3 shape, bool isPlanar, int outChannels)
{
    using BT = typename nvcv::cuda::BaseType<T>;

    const int ch = nvcv::cuda::NumElements<T>;

    nvcv::Tensor fg   = MakeCompositeTensor<BT>(shape, ch, isPlanar);
    nvcv::Tensor bg   = MakeCompositeTensor<BT>(shape, ch, isPlanar);
    nvcv::Tensor mask = MakeCompositeTensor<M>(shape, 1, isPlanar);
    nvcv::Tensor dst  = MakeCompositeTensor<BT>(shape, outChannels, isPlanar);

    benchutils::FillTensor<BT>(fg, benchutils::CheckerboardValues<BT>());
    benchutils::FillTensor<BT>(bg, benchutils::CheckerboardValues<BT>());
    benchutils::FillTensor<M>(mask, [](const long4_16a &) { return 1; });

    benchutils::warmup_and_exec(state, BENCH_COMPOSITE_WARMUP_ITERATIONS,
                                [&op, &fg, &bg, &mask, &dst](cudaStream_t s) { op(s, fg, bg, mask, dst); });
}

template<typename T, typename M>
void FillCompositeVarShapeInputs(nvcv::ImageBatchVarShape &fg, nvcv::ImageBatchVarShape &bg,
                                 nvcv::ImageBatchVarShape &mask, int3 shape, bool isPlanar)
{
    if (isPlanar)
    {
        benchutils::FillPlanarImageBatch<T>(fg, long2{shape.z, shape.y}, long2{0, 0});
        benchutils::FillPlanarImageBatch<T>(bg, long2{shape.z, shape.y}, long2{0, 0});
    }
    else
    {
        benchutils::FillImageBatch<T>(fg, long2{shape.z, shape.y}, long2{0, 0}, benchutils::CheckerboardValues<T>());
        benchutils::FillImageBatchLike<T>(bg, fg, benchutils::CheckerboardValues<T>());
    }

    benchutils::FillImageBatchLike<M>(mask, fg, [](const long4_16a &) { return 1; });
}

template<typename OutT>
void FillCompositeVarShapeOutput(nvcv::ImageBatchVarShape &dst, const nvcv::ImageBatchVarShape &fg, bool isPlanar)
{
    if (isPlanar)
    {
        benchutils::FillPlanarImageBatchLike<OutT>(dst, fg);
    }
    else
    {
        nvcv::ImageFormat format;
        if constexpr (nvcv::cuda::NumElements<OutT> == 3)
        {
            format = nvcv::FMT_RGB8;
        }
        else
        {
            static_assert(nvcv::cuda::NumElements<OutT> == 4);
            format = nvcv::FMT_RGBA8;
        }
        benchutils::FillImageBatchLike<OutT>(
            dst, fg, [](const long4_16a &) { return OutT{0}; }, format);
    }
}

template<typename T, typename M>
void RunVarShapeComposite(nvbench::state &state, cvcuda::Composite &op, int3 shape, bool isPlanar, int outChannels)
{
    nvcv::ImageBatchVarShape fg(shape.x);
    nvcv::ImageBatchVarShape bg(shape.x);
    nvcv::ImageBatchVarShape mask(shape.x);
    nvcv::ImageBatchVarShape dst(shape.x);

    FillCompositeVarShapeInputs<T, M>(fg, bg, mask, shape, isPlanar);
    if (outChannels == 3)
    {
        FillCompositeVarShapeOutput<uchar3>(dst, fg, isPlanar);
    }
    else
    {
        FillCompositeVarShapeOutput<uchar4>(dst, fg, isPlanar);
    }

    benchutils::warmup_and_exec(state, BENCH_COMPOSITE_WARMUP_ITERATIONS,
                                [&op, &fg, &bg, &mask, &dst](cudaStream_t s) { op(s, fg, bg, mask, dst); });
}

template<typename T, typename M>
bool DispatchCompositeRun(nvbench::state &state, cvcuda::Composite &op, int3 shape, benchutils::InputKind inputKind,
                          CompositeLayout layout, int outChannels)
{
    if (SkipUnsupportedCompositeCase(state, inputKind, layout))
    {
        return false;
    }

    if (IsFakePlanar(layout))
    {
        RunFakePlanarComposite<T, M>(state, op, shape, outChannels);
        return true;
    }

    if (inputKind == benchutils::InputKind::Tensor)
    {
        RunTensorComposite<T, M>(state, op, shape, IsPlanar(layout), outChannels);
        return true;
    }

    RunVarShapeComposite<T, M>(state, op, shape, IsPlanar(layout), outChannels);
    return true;
}

template<typename T, typename M = uint8_t>
inline void composite(nvbench::state &state, nvbench::type_list<T>)
try
{
    int3                        shape       = benchutils::GetShape<3, int3>(state.get_string("shape"));
    const benchutils::InputKind inputKind   = benchutils::GetInputKind(state.get_string("inputKind"));
    auto                        layout      = benchutils::GetStringParam(state, "layout", "NHWC");
    int                         outChannels = benchutils::GetIntParam<int>(state, "outChannels");

    CompositeLayout compositeLayout = ValidateCompositeBenchParams(state, layout, outChannels);
    if (compositeLayout == CompositeLayout::INVALID)
    {
        return;
    }

    using BT = typename nvcv::cuda::BaseType<T>;

    state.add_global_memory_reads(shape.x * shape.y * shape.z * (sizeof(T) * 2 + sizeof(M)));
    state.add_global_memory_writes(shape.x * shape.y * shape.z * outChannels * sizeof(BT));

    cvcuda::Composite op;
    DispatchCompositeRun<T, M>(state, op, shape, inputKind, compositeLayout, outChannels);
}
CVCUDA_BENCH_SKIP_ERRORS(state)

// Use auto-generated type list from bench_params.json

NVBENCH_BENCH_TYPES(composite, NVBENCH_TYPE_AXES(BENCH_COMPOSITE_TYPES))
BENCH_COMPOSITE_AXES;
